### PR TITLE
tcpinfo.proto

### PR DESCRIPTION
**Pull request by gfr10598**
When I made mlab/tcp_info public, it divorced it from the fork from which the previous PR was pulling.  No idea how to fix that, so I've generated a new pull request.

---

This change is [Reviewable](https://reviewable.io/reviews/m-lab/tcpinfo_lib/2)

### CODE REVIEW

File: tcpinfo.proto [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        INET_DIAG_MARK,   // NEW - should we include this???
+        INET_DIAG_BBRINFO,  // NEW Done.
+        __INET_DIAG_MAX,
+};
+
+*/
+
+message EndPoint {
+  optional int32 port = 1;
+  optional bytes ip = 2;  // 4 or 16 bytes for ipv4 or ipv6 address.
```
Comment (pboothe) [inline tcpinfo.proto:46]: If we want the IP to contain either v4 or v6, then it should be a submessage, right?

```
message IP {
  optional AddressFamily family = 1;
  optional bytes address = 2;  // 4 or 16 bytes for ipv4 or ipv6 address.
}
```

Putting the `AddressFamily` in the `InetDiagMsgProto` feels like separating info that should travel together.

File: tcpinfo.proto [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+  optional InetSocketIDProto sock_id = 5;
+  optional int32 expires =  6;
+  optional int32 rqueue  =  7;
+  optional int32 wqueue  =  8;
+  optional int32 uid     =  9;
+  optional int32 inode   = 10;
+}
+
+// Proto representation for struct tcpvegas_info.
+message TCPVegasInfoProto {
```
Comment (pboothe) [inline tcpinfo.proto:98]: I don't anticipate us ever needing this message type. We have no plans to ever deploy TCP-Vegas

File: tcpinfo.proto [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  optional int32 pacing_gain = 3;  // pacing gain shifted left 8 bits
+  optional int32 cwnd_gain = 4;  // cwnd gain shifted left 8 bits.
+}
+
+// This proto is intended to precisely represent the raw data from tcpinfo.
+// (Derived data will be represented in other protos.)
+message TCPInfoProto {
+  // state through rcv_wscale are 7 x _u8
+  // https://datatracker.ietf.org/doc/draft-ietf-tcpm-rfc793bis/
+  enum TCPState {  //  from tcp_states.h
```
Comment (pboothe) [inline tcpinfo.proto]: This is defined twice? Is it okay for them to ever be out of sync?

File: tcpinfo.proto [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+// We could optionally expand these into 6 bools.
+  optional int32  options = 6;
+  optional bool has_ts_opt = 601;
+  optional bool has_sack_opt = 602;
+  optional bool has_wscale_opt =  603;  // Or just use has_snd_wscale
+  optional bool has_ecn_opt = 604;
+  optional bool has_ecnseen_opt = 605;
+  optional bool has_fastopen_opt = 606;
+
+  optional bool delivery_rate_app_limited = 690;
```
Comment (pboothe) [inline tcpinfo.proto]: Question about field numbering asked in person

File: tcpinfo.proto (hunk 3 above)
Comment (gfr10598) [reply tcpinfo.proto]: Moved this up to a common definition outside the messages.

File: tcpinfo.proto (hunk 2 above)
Comment (gfr10598) [reply tcpinfo.proto:98]: Good to know.  I see no harm in leaving it though.

File: tcpinfo.proto (hunk 1 above)
Comment (gfr10598) [reply tcpinfo.proto:46]: Agree, but that is how the inet_diag.h structs group the fields.  In the proto, we have the advantage that we can infer the family from the number of bytes in the ip field, but the authority is the family field in the parent struct (or proto).  I've added an extended code comment discussing this.

File: tcpinfo.proto (hunk 4 above)
Comment (gfr10598) [reply tcpinfo.proto]: Fixed and commented.

**Review by pboothe — APPROVED** [review]
LGTM.

**Comment by pboothe** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/tcpinfo_lib/2#-:-KTz1NpYZigUOfpomVVg:bnfp4nl)_

**Pull request merged**